Ubc various updates

dquispeott opened the pull request:
1) added new method:
Optional<Course> updateCourse(Course course) throws IOException;
files changes: CourseImpl, CourseWriter, CourseManagerUTest, updateCourseSuccess.json, Course

2) updated method signature from:
public Optional<Enrollment> enrollUser(String courseId, String userId) throws InvalidOauthTokenException, IOException {
to:
public Optional<Enrollment> enrollUser(Enrollment enrollment) throws InvalidOauthTokenException, IOException {
Reason: we have a requirement to set user role and enrollment state at the time of enrollment
files changed: EnrollmentImpl,EnrollmentWriter, DropCourseUTest

3) added new classes to implement method:
public List<Role> listRoles(ListRolesOptions options) throws IOException
files changed: RolesImpl, RolesReader, RolesWriter, Role, ListRolesOption

4)
added new method:
public Optional<Section> createSection(String courseId, Section section) throws IOException {
files changed: SectionsImpl, SectionWriter, SectionManagerUTest, createSectionSuccess.json

5) 
added new method:
public Optional<User> showUserDetails(ShowUserDetailsOptions options) throws IOException{
files changes: UserImpl, UserReader, ShowUserDetailsOptions

6)
updated common class CanvasApiFactory to include RolesImpl and SectionsImpl

kawesome commented:
Can one of the admins verify this patch?

ToeBee commented:
Thanks, this is a hefty chunk of code!

I did just merge #53 that implemented Section creation/deletion before I took a good look at this PR. After a small fix, it handles the additional SIS reactivation option that you didn't implement. But on the other hand, you added unit tests! :) Can you do a merge from master or a rebase to resolve that conflict?

Some minor whitespace nits: Please use spaces instead of tabs. Personally, I'm not a fanatic either way but so far this project has used spaces so I prefer to keep it consistent. In fact, I bet you use Eclipse? I see some mixing of tabs and spaces on the same line (in `UserImpl`) that I've seen my own Eclipse doing before. I don't know why Eclipse thinks it is ok to mix tabs and spaces on the same line. Seems like all that does is tick off the space people, the tab people **and** the OCD people all at once.

dquispeott commented:
Hi Toby,

I rebased from master, merged changes, removed tabs, updated section test.

I am new to git, hope it is fine now. Please review and let me know if
anything else is needed?

How do other developers use spaces instead of tabs? Using eclipse I have to
manually review each line, which does not seem very productive.

Thanks



On Wed, Aug 16, 2017 at 4:57 PM, Toby Murray <notifications@github.com>
wrote:

> Thanks, this is a hefty chunk of code!
>
> I did just merge #53 <https://github.com/kstateome/canvas-api/pull/53>
> that implemented Section creation/deletion before I took a good look at
> this PR. After a small fix, it handles the additional SIS reactivation
> option that you didn't implement. But on the other hand, you added unit
> tests! :) Can you do a merge from master or a rebase to resolve that
> conflict?
>
> Some minor whitespace nits: Please use spaces instead of tabs. Personally,
> I'm not a fanatic either way but so far this project has used spaces so I
> prefer to keep it consistent. In fact, I bet you use Eclipse? I see some
> mixing of tabs and spaces on the same line (in UserImpl) that I've seen
> my own Eclipse doing before. I don't know why Eclipse thinks it is ok to
> mix tabs and spaces on the same line. Seems like all that does is tick off
> the space people, the tab people *and* the OCD people all at once.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kstateome/canvas-api/pull/54#issuecomment-322931300>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AaumTxtLRNS5vKmKmuijMp3Tf_eLbswFks5sY4HcgaJpZM4O22k_>
> .
>

ToeBee commented:
Ok, sorry for the delay here. I'm trying to handle pull requests on this project quickly but between the eclipse and the first week of classes, I have been a bit distracted. Thanks for the touch ups.

Eclipse has some settings to replace tabs with spaces. I think there is actually a file that could be checked into git that will make Eclipse automatically fix whitespace to conform to the project standards. I should probably look into that...

I will let Jenkins test this pull request now and see what it finds.

Jenkins: ok to test

ToeBee commented on src/main/java/edu/ksu/canvas/requestOptions/ListRolesOptions.java:
```diff
@@ -0,0 +1,26 @@
+ * See <a href="https://canvas.instructure.com/doc/api/roles.html#method.role_overrides.api_index">
+ *              https://canvas.instructure.com/doc/api/roles.html#method.role_overrides.api_index</a>
+ */
+public class ListRolesOptions extends BaseOptions {
```
This is an appropriate use of an options class for the `show_inherited` parameter. For bonus points, you could also implement the optional `state[]` parameter. This would involve creating an enum for the allowed values (there are only two in this case) and then accepting a list of that enum type. There are several similar methods already implemented. Like the `GetSingleCourseOptions.includes` method.

ToeBee commented on src/main/java/edu/ksu/canvas/interfaces/RolesReader.java:
```diff
@@ -0,0 +1,17 @@
+import java.io.IOException;
+import java.util.List;
+
+public interface RolesReader extends CanvasReader<Role, RolesReader> {
```
To match the naming of the rest of the interfaces in this package, this should probably be singular like `RoleReader` instead of `RolesReader`. I guess this applies to the `RolesImpl` class too (although I see the existing `SectionsImpl` class doesn't follow this convention)

dquispeott commented:
Hi Toby,

thanks for the feedback, updated, please review, have a great weekend!

David

On Thu, Aug 24, 2017 at 3:17 PM, Toby Murray <notifications@github.com>
wrote:

> *@ToeBee* requested changes on this pull request.
>
> Looks pretty good. I have added a couple of inline comments with a few
> more suggested tweaks.
> ------------------------------
>
> In src/main/java/edu/ksu/canvas/impl/UserImpl.java
> <https://github.com/kstateome/canvas-api/pull/54#discussion_r135141932>:
>
> > @@ -61,6 +66,24 @@ public UserImpl(String canvasBaseUrl, Integer apiVersion, OauthToken oauthToken,
>      }
>
>      @Override
> +    public Optional<User> showUserDetails(ShowUserDetailsOptions options) throws IOException{
>
> This ShowUserDetailsOptions class is only dealing with the ID of the
> object being queried. And it doesn't actually handle all the possible
> options either. I believe it can be any of:
>
>    - Numeric Canvas user ID
>    - A string like sis_integration_id:123 to query by SIS integration ID
>    - A string like sis_user_id:ABC to query by SIS ID
>    - The special string self that the Canvas API interprets as "the
>    currently logged in user"
>
> I think you can just change this method to accept a String and the calling
> code can decide what type of ID it wants to query by. This is how, for
> example, the QuizImpl.getSingleQuiz method does it. The Options classes
> are more intended to deal with optional parameters that some API calls take
> like various include[]=... options that get added to the URL as query
> parameters.
> ------------------------------
>
> In src/main/java/edu/ksu/canvas/requestOptions/ListRolesOptions.java
> <https://github.com/kstateome/canvas-api/pull/54#discussion_r135142282>:
>
> > @@ -0,0 +1,26 @@
> +package edu.ksu.canvas.requestOptions;
> +
> +/**
> + * A class to represent parameters to be passed to the "List the roles available to an account" API call.
> + * See <a href="https://canvas.instructure.com/doc/api/roles.html#method.role_overrides.api_index">
> + *              https://canvas.instructure.com/doc/api/roles.html#method.role_overrides.api_index</a>
> + */
> +public class ListRolesOptions extends BaseOptions {
>
> This is an appropriate use of an options class for the show_inherited
> parameter. For bonus points, you could also implement the optional state[]
> parameter. This would involve creating an enum for the allowed values
> (there are only two in this case) and then accepting a list of that enum
> type. There are several similar methods already implemented. Like the
> GetSingleCourseOptions.includes method.
> ------------------------------
>
> In src/main/java/edu/ksu/canvas/interfaces/RolesReader.java
> <https://github.com/kstateome/canvas-api/pull/54#discussion_r135144364>:
>
> > @@ -0,0 +1,17 @@
> +package edu.ksu.canvas.interfaces;
> +
> +import edu.ksu.canvas.model.Role;
> +import edu.ksu.canvas.requestOptions.ListRolesOptions;
> +
> +import java.io.IOException;
> +import java.util.List;
> +
> +public interface RolesReader extends CanvasReader<Role, RolesReader> {
>
> To match the naming of the rest of the interfaces in this package, this
> should probably be singular like RoleReader instead of RolesReader. I
> guess this applies to the RolesImpl class too (although I see the
> existing SectionsImpl class doesn't follow this convention)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kstateome/canvas-api/pull/54#pullrequestreview-58523080>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AaumT680k0qkE6f-RekrD5ej1F6jyTv-ks5sbfaCgaJpZM4O22k_>
> .
>

dquispeott commented:
Thanks Toby!

On Wed, Aug 30, 2017 at 1:28 PM, Toby Murray <notifications@github.com>
wrote:

> Merged #54 <https://github.com/kstateome/canvas-api/pull/54>.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kstateome/canvas-api/pull/54#event-1228369145>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AaumT2_wzNk6jAtHGF24qeO9v_rAoQjYks5sdcX1gaJpZM4O22k_>
> .
>
